# Exploratory Data Analysis — SkyWest Flight Delay Prediction
**ENSF 444 – Machine Learning Systems | Group 45**

This notebook explores the SkyWest Airlines flight dataset to understand data quality, feature distributions, and patterns that will guide our preprocessing and modeling decisions.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

pd.set_option('display.max_columns', None)

# Create plots directory
PLOTS_DIR = 'plots'
os.makedirs(PLOTS_DIR, exist_ok=True)

# Load dataset
df = pd.read_csv('../data/skywest_flights.csv', low_memory=False)
print(f'Shape: {df.shape[0]:,} rows × {df.shape[1]} columns')

Shape: 56,811 rows × 120 columns


## 1. Data Overview

In [2]:
df.head()

,Year,Quarter,Month,DayofMonth,DayOfWeek,FlightDate,Marketing_Airline_Network,Operated_or_Branded_Code_Share_Partners,DOT_ID_Marketing_Airline,IATA_Code_Marketing_Airline,Flight_Number_Marketing_Airline,Originally_Scheduled_Code_Share_Airline,DOT_ID_Originally_Scheduled_Code_Share_Airline,IATA_Code_Originally_Scheduled_Code_Share_Airline,Flight_Num_Originally_Scheduled_Code_Share_Airline,Operating_Airline,DOT_ID_Operating_Airline,IATA_Code_Operating_Airline,Tail_Number,Flight_Number_Operating_Airline,OriginAirportID,OriginAirportSeqID,OriginCityMarketID,Origin,OriginCityName,OriginState,OriginStateFips,OriginStateName,OriginWac,DestAirportID,DestAirportSeqID,DestCityMarketID,Dest,DestCityName,DestState,DestStateFips,DestStateName,DestWac,CRSDepTime,DepTime,DepDelay,DepDelayMinutes,DepDel15,DepartureDelayGroups,DepTimeBlk,TaxiOut,WheelsOff,WheelsOn,TaxiIn,CRSArrTime,ArrTime,ArrDelay,ArrDelayMinutes,ArrDel15,ArrivalDelayGroups,ArrTimeBlk,Cancelled,CancellationCode,Diverted,CRSElapsedTime,ActualElapsedTime,AirTime,Flights,Distance,DistanceGroup,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay,FirstDepTime,TotalAddGTime,LongestAddGTime,DivAirportLandings,DivReachedDest,DivActualElapsedTime,DivArrDelay,DivDistance,Div1Airport,Div1AirportID,Div1AirportSeqID,Div1WheelsOn,Div1TotalGTime,Div1LongestGTime,Div1WheelsOff,Div1TailNum,Div2Airport,Div2AirportID,Div2AirportSeqID,Div2WheelsOn,Div2TotalGTime,Div2LongestGTime,Div2WheelsOff,Div2TailNum,Div3Airport,Div3AirportID,Div3AirportSeqID,Div3WheelsOn,Div3TotalGTime,Div3LongestGTime,Div3WheelsOff,Div3TailNum,Div4Airport,Div4AirportID,Div4AirportSeqID,Div4WheelsOn,Div4TotalGTime,Div4LongestGTime,Div4WheelsOff,Div4TailNum,Div5Airport,Div5AirportID,Div5AirportSeqID,Div5WheelsOn,Div5TotalGTime,Div5LongestGTime,Div5WheelsOff,Div5TailNum,Duplicate,Unnamed: 119
0,2024,1,1,20,6,2024-01-20,UA,UA_CODESHARE,19977,UA,5464,NaN,NaN,NaN,NaN,OO,20304,OO,N783SK,5464,14489,1448903,34489,RDM,"Bend/Redmond, OR",OR,41,Oregon,92,14771,1477104,32457,SFO,"San Francisco, CA",CA,6,California,91,545,NaN,NaN,NaN,NaN,NaN,0001-0559,NaN,NaN,NaN,NaN,745,NaN,NaN,NaN,NaN,NaN,0700-0759,1.0,B,0.0,120.0,NaN,NaN,1.0,462.0,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN
1,2024,1,1,20,6,2024-01-20,UA,UA_CODESHARE,19977,UA,5612,NaN,NaN,NaN,NaN,OO,20304,OO,N107SY,5612,14489,1448903,34489,RDM,"Bend/Redmond, OR",OR,41,Oregon,92,12892,1289208,32575,LAX,"Los Angeles, CA",CA,6,California,91,1940,NaN,NaN,NaN,NaN,NaN,1900-1959,NaN,NaN,NaN,NaN,2155,NaN,NaN,NaN,NaN,NaN,2100-2159,1.0,B,0.0,135.0,NaN,NaN,1.0,726.0,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN
2,2024,1,1,20,6,2024-01-20,UA,UA_CODESHARE,19977,UA,5083,NaN,NaN,NaN,NaN,OO,20304,OO,N959SW,5083,14237,1423706,34237,PRC,"Prescott, AZ",AZ,4,Arizona,81,11292,1129202,30325,DEN,"Denver, CO",CO,8,Colorado,82,931,923.0,-8.0,0.0,0.0,-1.0,0900-0959,8.0,931.0,1050.0,10.0,1133,1100.0,-33.0,0.0,0.0,-2.0,1100-1159,0.0,NaN,0.0,122.0,97.0,79.0,1.0,557.0,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN
3,2024,1,1,20,6,2024-01-20,UA,UA_CODESHARE,19977,UA,5175,NaN,NaN,NaN,NaN,OO,20304,OO,N959SW,5175,11292,1129202,30325,DEN,"Denver, CO",CO,8,Colorado,82,11865,1186503,31865,GCC,"Gillette, WY",WY,56,Wyoming,88,1220,1214.0,-6.0,0.0,0.0,-1.0,1200-1259,14.0,1228.0,1318.0,7.0,1342,1325.0,-17.0,0.0,0.0,-2.0,1300-1359,0.0,NaN,0.0,82.0,71.0,50.0,1.0,313.0,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56811 entries, 0 to 56810
Columns: 120 entries, Year to Unnamed: 119
dtypes: float64(74), int64(23), object(23)
memory usage: 52.0+ MB


In [4]:
df.describe()

,Year,Quarter,Month,DayofMonth,DayOfWeek,DOT_ID_Marketing_Airline,Flight_Number_Marketing_Airline,Originally_Scheduled_Code_Share_Airline,DOT_ID_Originally_Scheduled_Code_Share_Airline,IATA_Code_Originally_Scheduled_Code_Share_Airline,Flight_Num_Originally_Scheduled_Code_Share_Airline,DOT_ID_Operating_Airline,Flight_Number_Operating_Airline,OriginAirportID,OriginAirportSeqID,OriginCityMarketID,OriginStateFips,OriginWac,DestAirportID,DestAirportSeqID,DestCityMarketID,DestStateFips,DestWac,CRSDepTime,DepTime,DepDelay,DepDelayMinutes,DepDel15,DepartureDelayGroups,TaxiOut,WheelsOff,WheelsOn,TaxiIn,CRSArrTime,ArrTime,ArrDelay,ArrDelayMinutes,ArrDel15,ArrivalDelayGroups,Cancelled,Diverted,CRSElapsedTime,ActualElapsedTime,AirTime,Flights,Distance,DistanceGroup,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay,FirstDepTime,TotalAddGTime,LongestAddGTime,DivAirportLandings,DivReachedDest,DivActualElapsedTime,DivArrDelay,DivDistance,Div1AirportID,Div1AirportSeqID,Div1WheelsOn,Div1TotalGTime,Div1LongestGTime,Div1WheelsOff,Div2AirportID,Div2AirportSeqID,Div2WheelsOn,Div2TotalGTime,Div2LongestGTime,Div2WheelsOff,Div3Airport,Div3AirportID,Div3AirportSeqID,Div3WheelsOn,Div3TotalGTime,Div3LongestGTime,Div3WheelsOff,Div3TailNum,Div4Airport,Div4AirportID,Div4AirportSeqID,Div4WheelsOn,Div4TotalGTime,Div4LongestGTime,Div4WheelsOff,Div4TailNum,Div5Airport,Div5AirportID,Div5AirportSeqID,Div5WheelsOn,Div5TotalGTime,Div5LongestGTime,Div5WheelsOff,Div5TailNum,Unnamed: 119
count,56811.0,56811.0,56811.0,56811.000000,56811.000000,56811.000000,56811.000000,0.0,0.0,0.0,0.0,56811.0,56811.000000,56811.000000,5.681100e+04,56811.000000,56811.000000,56811.000000,56811.000000,5.681100e+04,56811.000000,56811.000000,56811.000000,56811.000000,54541.000000,54476.000000,54476.000000,54476.000000,54476.000000,54476.000000,54476.000000,54349.000000,54349.000000,56811.000000,54349.000000,54206.000000,54206.000000,54206.000000,54206.000000,56811.000000,56811.000000,56811.000000,54206.000000,54206.000000,56811.0,56811.000000,56811.000000,13082.000000,13082.000000,13082.000000,13082.000000,13082.000000,572.000000,572.000000,572.000000,56811.000000,224.000000,143.000000,143.000000,224.000000,270.000000,2.700000e+02,270.000000,270.000000,270.000000,145.000000,6.000000,6.000000e+00,6.000000,6.000000,6.000000,4.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
mean,2024.0,1.0,1.0,16.092464,3.796289,19880.691433,4395.790974,NaN,NaN,NaN,NaN,20304.0,4395.790974,13045.297090,1.304534e+06,31973.855046,23.979089,72.863284,13044.999824,1.304504e+06,31973.996568,23.981078,72.857334,1328.451145,1345.433692,19.802078,23.601788,0.217197,0.419910,21.865978,1380.349769,1485.540231,8.057002,1494.573533,1493.648752,15.202837,24.286629,0.241339,0.224200,0.041911,0.003943,107.669976,103.374719,73.481312,1.0,459.047174,2.341677,44.477450,21.655251,11.393136,0.179865,19.051980,1264.765734,51.057692,50.265734,0.011318,0.638393,497.447552,423.461538,49.343750,12699.037037,1.269907e+06,1501.266667,21.770370,17.711111,1340.744828,12894.333333,1.289436e+06,1313.000000,26.666667,20.500000,1142.750000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,0.0,0.0,0.0,8.903721,2.005500,87.284107,857.332261,NaN,NaN,NaN,NaN,0.0,857.332261,1584.730353,1.584730e+05,1491.252013,17.728021,19.798407,1584.700086,1.584700e+05,1491.106133,17.729619,19.805222,459.860856,466.304751,81.788592,80.591896,0.412341,2.991509,12.680506,464.728136,476.821200,6.490272,468.761976,479.507303,83.440321,80.081122,0.427899,3.126218,0.200387,0.062669,37.751806,40.099604,36.901712,0.0,282.624839,1.172401,108.823325,83.857066,32.557642,4.723627,69.425677,503.226717,36.819318,35.557644,0.263943,0.481542,339.908473,367.231530,87.386532,1464.136971,1.464140e+05,480.481306,20.140193,17.090125,388.210778,1394.582614,1.394594e+05,212.605738,21.860162,14.896308,225.013148,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN

In [5]:
# List all columns with their dtypes
pd.DataFrame({'dtype': df.dtypes, 'non_null': df.notnull().sum(), 'null_pct': (df.isnull().sum() / len(df) * 100).round(1)}).sort_values('null_pct', ascending=False)

,dtype,non_null,null_pct
Unnamed: 119,float64,0,100.0
Div3AirportID,float64,0,100.0
Div4Airport,float64,0,100.0
Div3TailNum,float64,0,100.0
Div3WheelsOff,float64,0,100.0
...,...,...,...
Dest,object,56811,0.0
DestCityMarketID,int64,56811,0.0
DestAirportSeqID,int64,56811,0.0
DestAirportID,int64,56811,0.0


## 2. Missing Value Analysis

In [6]:
# Missing values — only columns with >0% missing
missing_pct = (df.isnull().sum() / len(df) * 100).round(1)
missing_pct = missing_pct[missing_pct > 0].sort_values(ascending=True)

# Color by severity: orange = partial, red = ≥95%
colors = ['#c44e52' if v >= 95 else '#dd8452' for v in missing_pct.values]

fig, ax = plt.subplots(figsize=(10, 22))
ax.barh(range(len(missing_pct)), missing_pct.values, color=colors)
ax.set_yticks(range(len(missing_pct)))
ax.set_yticklabels(missing_pct.index, fontsize=8)
ax.set_xlabel('Missing (%)')
ax.set_title(f'Missing Data by Feature ({len(missing_pct)} of {df.shape[1]} columns)')

# Legend
from matplotlib.patches import Patch
legend_elements = [Patch(facecolor='#dd8452', label='Partial missing'),
                   Patch(facecolor='#c44e52', label='≥95% missing')]
ax.legend(handles=legend_elements, loc='lower right')

plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/missing_data.png', dpi=150)
plt.close(fig)

n_zero = df.shape[1] - len(missing_pct)
n_partial = ((missing_pct > 0) & (missing_pct < 95)).sum()
n_drop = (missing_pct >= 95).sum()
print(f'0% missing (not shown): {n_zero} columns')
print(f'Partial missing (0-95%): {n_partial} columns')
print(f'≥95% missing (to drop): {n_drop} columns')

0% missing (not shown): 46 columns
Partial missing (0-95%): 21 columns
≥95% missing (to drop): 53 columns


## 3. Target Variable: Arrival Delay

In [7]:
# Drop rows where ArrDelay is missing (cancelled/diverted flights)
print(f'Rows before dropping missing ArrDelay: {len(df):,}')
df_valid = df.dropna(subset=['ArrDelay']).copy()
print(f'Rows after: {len(df_valid):,} (dropped {len(df) - len(df_valid):,})')

# Engineer binary target
df_valid['Delayed'] = (df_valid['ArrDelay'] >= 15).astype(int)

# Class distribution
counts = df_valid['Delayed'].value_counts()
pcts = df_valid['Delayed'].value_counts(normalize=True) * 100

print(f'\nClass Distribution:')
print(f'  On-Time (0): {counts[0]:,} ({pcts[0]:.1f}%)')
print(f'  Delayed (1): {counts[1]:,} ({pcts[1]:.1f}%)')
print(f'  Imbalance ratio: {counts[0]/counts[1]:.1f}:1')

Rows before dropping missing ArrDelay: 56,811
Rows after: 54,206 (dropped 2,605)

Class Distribution:
  On-Time (0): 41,124 (75.9%)
  Delayed (1): 13,082 (24.1%)
  Imbalance ratio: 3.1:1


In [8]:
# Target class distribution and raw ArrDelay histogram
fig, axes = plt.subplots(1, 2, figsize=(12, 5))

# Class distribution bar chart
labels = ['On-Time', 'Delayed (≥15 min)']
colors = ['#4c72b0', '#dd8452']
axes[0].bar(labels, [counts[0], counts[1]], color=colors)
axes[0].set_title('Target Class Distribution')
axes[0].set_ylabel('Count')
for i, v in enumerate([counts[0], counts[1]]):
    axes[0].text(i, v + 500, f'{v:,}', ha='center', fontsize=11)

# Raw ArrDelay distribution
axes[1].hist(df_valid['ArrDelay'].clip(-60, 180), bins=80, color='steelblue', edgecolor='white')
axes[1].axvline(x=15, color='red', linestyle='--', linewidth=2, label='15-min threshold')
axes[1].set_xlabel('Arrival Delay (minutes)')
axes[1].set_ylabel('Count')
axes[1].set_title('Arrival Delay Distribution (clipped to [-60, 180])')
axes[1].legend()

plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/target_distribution.png', dpi=150)
plt.close(fig)

## 4. Feature Distributions

In [9]:
# Numerical feature distributions
num_features = ['CRSDepTime', 'CRSArrTime', 'CRSElapsedTime', 'Distance', 'DepDelay']

fig, axes = plt.subplots(2, 3, figsize=(16, 8))
axes = axes.flatten()

for i, col in enumerate(num_features):
    data = df_valid[col].dropna()
    axes[i].hist(data, bins=50, color='steelblue', edgecolor='white')
    axes[i].set_title(f'{col}\n(mean={data.mean():.1f}, std={data.std():.1f})')
    axes[i].set_xlabel(col)
    axes[i].set_ylabel('Count')

axes[-1].axis('off')
plt.suptitle('Numerical Feature Distributions', fontsize=14)
plt.tight_layout(rect=[0, 0, 1, 0.95])
fig.savefig(f'{PLOTS_DIR}/numerical_distributions.png', dpi=150)
plt.close(fig)

In [10]:
# DepDelay vs ArrDelay scatter — visualizing the data leakage concern
fig, ax = plt.subplots(figsize=(8, 8))
sample = df_valid.sample(n=5000, random_state=42)
ax.scatter(sample['DepDelay'], sample['ArrDelay'], alpha=0.3, s=10, color='steelblue')
ax.axhline(y=15, color='red', linestyle='--', linewidth=1.5, label='ArrDelay = 15 min threshold')
ax.axvline(x=15, color='#dd8452', linestyle='--', linewidth=1.5, label='DepDelay = 15 min')
ax.set_xlabel('Departure Delay (min)')
ax.set_ylabel('Arrival Delay (min)')
ax.set_title(f'DepDelay vs ArrDelay (r = {df_valid["DepDelay"].corr(df_valid["ArrDelay"]):.3f})')
ax.set_xlim(-30, 200)
ax.set_ylim(-60, 200)
ax.legend()
plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/depdelay_vs_arrdelay.png', dpi=150)
plt.close(fig)

In [11]:
# Feature distributions by delay status — DepDelay and Distance
on_time = df_valid[df_valid['Delayed'] == 0]
delayed = df_valid[df_valid['Delayed'] == 1]

fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# DepDelay by delay status
axes[0].hist(on_time['DepDelay'].clip(-30, 120), bins=60, alpha=0.6, color='#4c72b0', label='On-Time', edgecolor='white')
axes[0].hist(delayed['DepDelay'].clip(-30, 120), bins=60, alpha=0.6, color='#dd8452', label='Delayed', edgecolor='white')
axes[0].set_xlabel('Departure Delay (min)')
axes[0].set_ylabel('Count')
axes[0].set_title('DepDelay Distribution by Delay Status')
axes[0].legend()

# Distance by delay status
axes[1].hist(on_time['Distance'], bins=50, alpha=0.6, color='#4c72b0', label='On-Time', edgecolor='white')
axes[1].hist(delayed['Distance'], bins=50, alpha=0.6, color='#dd8452', label='Delayed', edgecolor='white')
axes[1].set_xlabel('Distance (miles)')
axes[1].set_ylabel('Count')
axes[1].set_title('Distance Distribution by Delay Status')
axes[1].legend()

plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/features_by_delay_status.png', dpi=150)
plt.close(fig)

In [12]:
# Flights by Day of Week (only 1 month of data, so month chart is not useful)
day_labels = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
day_counts = df_valid['DayOfWeek'].value_counts().sort_index()

fig, ax = plt.subplots(figsize=(8, 5))
ax.bar(day_labels, day_counts.values, color='steelblue')
ax.set_title('Flights by Day of Week')
ax.set_ylabel('Count')
plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/flights_by_day.png', dpi=150)
plt.close(fig)

## 5. Correlation Analysis

In [13]:
# Correlation heatmap of key numerical features
corr_cols = ['DepDelay', 'ArrDelay', 'CRSDepTime', 'CRSArrTime', 'CRSElapsedTime',
             'Distance', 'DayOfWeek', 'DayofMonth', 'TaxiOut', 'TaxiIn', 'AirTime']
corr_cols = [c for c in corr_cols if c in df_valid.columns]

corr_matrix = df_valid[corr_cols].corr()

fig, ax = plt.subplots(figsize=(12, 10))
sns.heatmap(corr_matrix, annot=True, fmt='.2f', cmap='coolwarm', center=0,
            square=True, linewidths=0.5, ax=ax)
ax.set_title('Correlation Matrix — Key Numerical Features')
plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/correlation_heatmap.png', dpi=150)
plt.close(fig)

print(f"DepDelay ↔ ArrDelay correlation: {df_valid['DepDelay'].corr(df_valid['ArrDelay']):.3f}")
print('→ Very high correlation confirms potential data leakage concern from proposal.')

DepDelay ↔ ArrDelay correlation: 0.982
→ Very high correlation confirms potential data leakage concern from proposal.


## 6. Temporal Patterns

In [14]:
# Delay rate by day of week and hour of day (month excluded — dataset only has January)
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

day_labels = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']

# By Day of Week
delay_by_day = df_valid.groupby('DayOfWeek')['Delayed'].mean() * 100
axes[0].bar(day_labels, delay_by_day.values, color='#dd8452')
axes[0].set_title('Delay Rate by Day of Week')
axes[0].set_ylabel('Delay Rate (%)')
axes[0].set_ylim(0, delay_by_day.max() * 1.3)
for i, v in enumerate(delay_by_day.values):
    axes[0].text(i, v + 0.5, f'{v:.1f}%', ha='center', fontsize=10)

# By Hour of Day
df_valid['DepHour'] = df_valid['CRSDepTime'] // 100
delay_by_hour = df_valid.groupby('DepHour')['Delayed'].mean() * 100
axes[1].plot(delay_by_hour.index, delay_by_hour.values, color='#dd8452', marker='o', linewidth=2)
axes[1].set_title('Delay Rate by Scheduled Departure Hour')
axes[1].set_xlabel('Hour of Day')
axes[1].set_ylabel('Delay Rate (%)')
axes[1].set_xticks(range(0, 24))
axes[1].tick_params(axis='x', rotation=45)

plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/delay_rate_temporal.png', dpi=150)
plt.close(fig)

## 7. Route & Airport Analysis

In [15]:
# Top 15 busiest origin and destination airports
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

top_origins = df_valid['Origin'].value_counts().head(15)
top_origins.plot(kind='barh', ax=axes[0], color='steelblue')
axes[0].set_title('Top 15 Origin Airports by Flight Count')
axes[0].set_xlabel('Number of Flights')
axes[0].invert_yaxis()

top_dests = df_valid['Dest'].value_counts().head(15)
top_dests.plot(kind='barh', ax=axes[1], color='steelblue')
axes[1].set_title('Top 15 Destination Airports by Flight Count')
axes[1].set_xlabel('Number of Flights')
axes[1].invert_yaxis()

plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/top_airports.png', dpi=150)
plt.close(fig)

In [16]:
# Airports with highest delay rates (minimum 100 flights)
origin_delay = df_valid.groupby('Origin').agg(
    flights=('Delayed', 'count'),
    delay_rate=('Delayed', 'mean')
).query('flights >= 100').sort_values('delay_rate', ascending=False)
origin_delay['delay_rate'] *= 100

fig, axes = plt.subplots(1, 2, figsize=(16, 6))

top_delayed = origin_delay.head(15)
axes[0].barh(top_delayed.index, top_delayed['delay_rate'], color='#dd8452')
axes[0].set_title('Top 15 Most Delayed Origin Airports (min 100 flights)')
axes[0].set_xlabel('Delay Rate (%)')
axes[0].invert_yaxis()
for i, (v, n) in enumerate(zip(top_delayed['delay_rate'], top_delayed['flights'])):
    axes[0].text(v + 0.3, i, f'{v:.1f}% (n={n:,})', va='center', fontsize=9)

least_delayed = origin_delay.tail(15)
axes[1].barh(least_delayed.index, least_delayed['delay_rate'], color='#4c72b0')
axes[1].set_title('Top 15 Most Punctual Origin Airports (min 100 flights)')
axes[1].set_xlabel('Delay Rate (%)')
axes[1].invert_yaxis()
for i, (v, n) in enumerate(zip(least_delayed['delay_rate'], least_delayed['flights'])):
    axes[1].text(v + 0.3, i, f'{v:.1f}% (n={n:,})', va='center', fontsize=9)

plt.tight_layout()
fig.savefig(f'{PLOTS_DIR}/airport_delay_rates.png', dpi=150)
plt.close(fig)

## 8. Key Takeaways

1. **Class Imbalance** — 75.9% on-time vs 24.1% delayed (3.1:1 ratio). Will need `class_weight='balanced'` or SMOTE.

2. **Data Leakage** — `DepDelay` ↔ `ArrDelay` correlation is 0.98. Models must be trained both with and without `DepDelay`.

3. **Droppable Columns** — 50+ columns are ≥95% null (diversion fields, codeshare metadata, `Unnamed: 119`). Safe to drop.

4. **Single-Month Dataset** — The SkyWest subset only contains January 2024 data, so month and year are not useful features.

5. **Temporal Patterns** — Delay rates increase throughout the day (cascading delays), peaking in evenings. Friday has the highest delay rate by day of week.

6. **Airport Effects** — Clear variation in delay rates across airports (7.7% at BFL vs 46.5% at MBS), making `Origin` and `Dest` useful features.

7. **Feature Selection** — Key candidates: `DayOfWeek`, `DayofMonth`, `CRSDepTime`, `CRSElapsedTime`, `Distance`, `Origin`, `Dest`, and optionally `DepDelay`.